### ISSUE TITLE
Bug in polyder

### ISSUE THREAD

**Issue by aytekinar** [Keno/Polynomials.jl#80]
### Description

Right now the only check in the package is `T<:Number`. Hence, we have `NaN`s and `Inf`s resulting in weird behaviours. One example was in #78, where the polynomial evaluation was wrong.

We also have a strange behaviour when taking the derivatives.

### Observed Behaviour

```julia
julia> using Polynomials
julia> p = Poly([NaN, 1, Inf])
Poly(NaN + 1.0⋅x + Inf⋅x^2)
julia> p'
Poly(1.0 + Inf⋅x)
julia> p''
Poly(Inf)
julia> p'''
Poly(0.0)
julia> polyder(polyint(p))
Poly(NaN + 1.0⋅x + Inf⋅x^2)
julia> polyint(polyder(p))
Poly(1.0⋅x + Inf⋅x^2)
```

### Desired Behaviour

If the package continues to support `NaN`s and `Inf`s, I believe any operation on `NaN`s should produce `NaN`, including the derivative. Right now, derivative of `p(x) = NaN` is `0.0`. Similarly, when `p(x) = Inf`, the derivative evaluates to `0.0`, again. However, when we apply the definition (at any point), we get `(Inf-Inf)/h`. Numerator, there, is also `NaN`. I assume, derivative of `Inf` should yield `NaN`, instead.

So, my proposed solution is at least as follows:

```julia
julia> using Polynomials
julia> p = Poly([NaN, 1, Inf])
Poly(NaN + 1.0⋅x + Inf⋅x^2)
julia> p'
Poly(NaN + Inf⋅x)
julia> p''
Poly(NaN)
julia> p'''
Poly(NaN)
julia> polyder(polyint(p))
Poly(NaN + 1.0⋅x + Inf⋅x^2)
julia> polyint(polyder(p))
Poly(NaN⋅x + Inf⋅x^2)
```

What do you think? Am I missing something?

**Note.** A PR will follow.

**Comment by jverzani** [comment]
Interesting questions about what to do with NaN and Inf. Float64 doesn't quite form a ring (no associativity, no additive inverse for NaN, ...), but perhaps it still makes sense to follow the definition of a formal derivative (https://en.wikipedia.org/wiki/Formal_derivative) for a commutative ring. This seems to imply the current behavior is the proper, though I see why it can be considered not to be so. 

In SymPy, the polynomial NaN + g(x) is just NaN, which makes its own sense. In that case the derivative of Poly(NaN) should be Poly(NaN)?

**Comment by aytekinar** [comment]
In my opinion, `NaN` should at least be preserved all the way down. Since the polynomial is something like `p(x) = NaN + x + ...` and referring to [MATLAB Documentation](https://se.mathworks.com/help/matlab/ref/nan.html) (and I am not claiming MATLAB is always right), any polynomial including `NaN`s should produce `Poly(NaN)` as a result (and hence, not the above suggestion of mine, either).

Then there are two good options:

1. In the constructor, add a check such as `any(isnan(x)) && new{T}([NaN], var)`, or,
2. Allow for usual construction, but in `polyval`, `polyder` and `polyint`, return always a `Poly(NaN)`.

As for (1) above, we again have a problem. Since, in Julia, we can resize 1-dimensional arrays and/or change their elements even in `immutable` types, we cannot prevent users from doing `p.a[1] = NaN` even though we catch this ill-posed behaviour in construction. Then, it is the question of whether we should allow for `NaN` constructions as in the case of `[NaN, 1, Inf, -Inf]` (and I do also know that the users shall not interfere with the implementation details as in the above example).

Maybe, Polynomials should also implement (overload) `isnan` and `isinf` methods from `Base`.

**Comment by jverzani** [comment]
All good points, I would think 2. is the way to go, but am curious what others think, as 1. is cleaner and works for well-behaved users.

**Comment by aytekinar** [comment]
A side note on derivation...

Wolfram Alpha seems to agree with the PR I have provided when it comes to [derivation](https://www.wolframalpha.com/input/?i=derivative+of+(Inf*x+%2B+Inf*x%5E2)). It also gives a similar result in [integration](https://www.wolframalpha.com/input/?i=integral+(Inf*x+%2B+Inf*x%5E2)+dx).

However, when you plug in a constant `NaN` term, it neglects `NaN` as if it were not there at all, and gives a different result when you try to take the integral or the derivative of the very same expressions.

Maybe we can stick to SymPy behaviour you mentioned before, and treat the `Inf` values as constant objects.

I would also opt for 2. above --- for me, if we cannot guarantee type invariance (in this case, with the current implementation of Julia, we cannot guarantee `NaN` invariance as we discussed), we should learn to live with it and take precautions accordingly.
